homeassistant/components/rituals_perfume_genie/sensor.py

```diff
@@ -0,0 +1,147 @@
+"""Support for Rituals Perfume Genie sensors."""
+from homeassistant.const import (
+    ATTR_BATTERY_CHARGING,
+    DEVICE_CLASS_BATTERY,
+    DEVICE_CLASS_SIGNAL_STRENGTH,
+)
+
+from .const import COORDINATORS, DEVICES, DOMAIN, HUB
+from .entity import SENSORS, DiffuserEntity
+
+ID = "id"
+TITLE = "title"
+ICON = "icon"
+WIFI = "wific"
+BATTERY = "battc"
+PERFUME = "rfidc"
+FILL = "fillc"
+
+BATTERY_CHARGING_ID = 21
+PERFUME_NO_CARTRIDGE_ID = 19
+FILL_NO_CARTRIDGE_ID = 12
+
+BATTERY_SUFFIX = " Battery"
+PERFUME_SUFFIX = " Perfume"
+FILL_SUFFIX = " Fill"
+WIFI_SUFFIX = " Wifi"
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the diffuser sensors."""
+    diffusers = hass.data[DOMAIN][config_entry.entry_id][DEVICES]
+    coordinators = hass.data[DOMAIN][config_entry.entry_id][COORDINATORS]
+    entities = []
+    for hublot, diffuser in diffusers.items():
+        coordinator = coordinators[hublot]
+        entities.append(DiffuserPerfumeSensor(diffuser, coordinator))
+        entities.append(DiffuserFillSensor(diffuser, coordinator))
+        entities.append(DiffuserWifiSensor(diffuser, coordinator))
+        if BATTERY in diffuser.data[HUB][SENSORS]:
+            entities.append(DiffuserBatterySensor(diffuser, coordinator))
+
+    async_add_entities(entities)
+
+
+class DiffuserPerfumeSensor(DiffuserEntity):
+    """Representation of a diffuser perfume sensor."""
+
+    def __init__(self, diffuser, coordinator):
+        """Initialize the perfume sensor."""
+        super().__init__(diffuser, coordinator, PERFUME_SUFFIX)
+
+    @property
+    def icon(self):
+        """Return the perfume sensor icon."""
+        if self.coordinator.data[HUB][SENSORS][PERFUME][ID] == PERFUME_NO_CARTRIDGE_ID:
+            return "mdi:tag-remove"
+        return "mdi:tag-text"
+
+    @property
+    def state(self):
+        """Return the state of the perfume sensor."""
+        return self.coordinator.data[HUB][SENSORS][PERFUME][TITLE]
+
+
+class DiffuserFillSensor(DiffuserEntity):
+    """Representation of a diffuser fill sensor."""
+
+    def __init__(self, diffuser, coordinator):
+        """Initialize the fill sensor."""
+        super().__init__(diffuser, coordinator, FILL_SUFFIX)
+
+    @property
+    def icon(self):
+        """Return the fill sensor icon."""
+        if self.coordinator.data[HUB][SENSORS][FILL][ID] == FILL_NO_CARTRIDGE_ID:
+            return "mdi:beaker-question"
+        return "mdi:beaker"
+
+    @property
+    def state(self):
+        """Return the state of the fill sensor."""
+        return self.coordinator.data[HUB][SENSORS][FILL][TITLE]
+
+
+class DiffuserBatterySensor(DiffuserEntity):
+    """Representation of a diffuser battery sensor."""
+
+    def __init__(self, diffuser, coordinator):
+        """Initialize the battery sensor."""
+        super().__init__(diffuser, coordinator, BATTERY_SUFFIX)
+
+    @property
```

Comment:
`icon` and `device_class` are mutually exclusive

https://developers.home-assistant.io/docs/core/entity/#advanced-properties

Battery iconography should be handled on the frontend

Reply:
Ideally the underlying package will do the conversion and present a percentage so its abstracted away from home assistant